Add function to estimate perspective transform between two planes

animeshsasan opened the pull request:
Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
In multi-session or multi-video experiments of the same setup, small shifts in the camera or setup can cause coordinate systems to differ between sessions. This prevents direct comparison or aggregation of tracked positional data across sessions. Even minor changes in angle, distance, or tilt can introduce perspective distortions that make one-to-one point correspondence unreliable without geometric alignment.

**What does this PR do?**
This PR adds functionality to compute homography transformations between sessions. This serves as the foundational step for aligning coordinate systems across recordings. A subsequent PR will integrate this functionality into the RegionOfInterest (RoI) class.

Key details:

- Implements a homography-based alignment suitable for planar scenes with perspective effects (where the camera is tilted, moved closer/further, or rotated, but the scene remains planar).
- Ensures at least four non-collinear, non-degenerate points are used to compute the transformation reliably.


## References
https://github.com/neuroinformatics-unit/movement/issues/565

## How has this PR been tested?
This code was tested using multiple dummy problems in local. Some of the toy problems are also available in the unit tests.

## Is this a breaking change?
No.

## Does this PR require an update to the documentation?
Yes.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

niksirbi commented on movement/transforms.py:
```diff
@@ -113,3 +116,155 @@ def scale(
+    srcPoints : np.ndarray
+        An array of shape (N, 2) representing N source points
+        in 2-dimensional space. N >= 4.
+        The data type is assumed to be float32.
```
I wouldn't saay "The data type is assumed to be float32". It's rather that your code coerces it into float32. I would probably just omit this sentence (for both arguments).

niksirbi commented on movement/transforms.py:
```diff
@@ -113,3 +116,155 @@ def scale(
     return scaled_data
+
+
+def get_homography_transform(
```
I wonder whether this should be named `compute_homography_transform`, to better match function names in our other modules.

niksirbi commented on movement/transforms.py:
```diff
@@ -113,3 +116,155 @@ def scale(
+
+    dim = srcPoints.shape[1]
+    if dim != 2:
+        raise ValueError("Points must be 2-dimensional.")
```
this is a very minor comment, but maybe we can refactor out the 3 shape-related validations to a separate func called `_validate_points_shape`?

niksirbi commented on movement/transforms.py:
```diff
@@ -113,3 +116,155 @@ def scale(
+):
+    assert len(src_pts) == len(dst_pts)
+    src_pts = np.asarray(src_pts)
+    dst_pts = np.asarray(dst_pts)
```
Are these checks necessary here? If `filter_invalid_points` is only ever going to be called by `get_homography_transform`, the datatype and equality of points are already guaranteed. We would need these checks only if other (future?) functions also call on `filter_invalid_points`.

niksirbi commented on movement/transforms.py:
```diff
@@ -113,3 +116,155 @@ def scale(
+    )
+
+
+def _is_collinear_set(points: np.ndarray, eps: float = 1e-6):
```
Similarly, I would add a one-line docstring stating that this uses am SVD-based rank check to determine if a set of points is collinear.

niksirbi reviewed (commented):
This looks mostly good to me @animeshsasan!

I've left some comments, but most of them are optional. Use your best judgement and don't hesitate to push back if necessary.

In the interest of keeping PRs [atomic](https://medium.com/@chikim79/the-art-of-creating-atomic-prs-16d13b3828e3) and quick-to-review, what do you think of wrapping up this PR with just the `findHomography` wrapper (after addressing my comments) and leaving the ROI API for another PR?

In that case,  you would have to update this PR's title and description accordingly.

animeshsasan commented:
Hey @niksirbi,
I’ve addressed some of the comments and replied to the others. I’ve also updated the PR description as suggested, and I’ll create a separate PR for the ROI changes.

niksirbi commented on movement/transforms.py:
```diff
@@ -113,3 +116,155 @@ def scale(
+
+
+def _filter_invalid_points(
+    src_pts: np.ndarray, dst_pts: np.ndarray, eps: float = 1e-6
```
Hmm, let me think through this.

Currently this `eps` value is only exposed at the level of the private `_filter_invalid_points`, `_is_collinear_three_` and `_is_collinear_set` functions, which are anyway not accessed by users. If we wanted to trully make it a user-controlled parameter, we'd have to expose it in `compute_homography_matrix`, or even at the ROI API level later.

You are right that we don't always know the scale, but I suspect in most cases users will be defining ROIs in the pixel coordinate system (as they are easiest to draw on top of an images/video). If two points are only a tiny fraction of a pixel away from each other, I think it's safe to say they are identical.

I would say to hard-code `eps` for now and see if anyone complains about it in the future, unless you found the results to be sensitibe to `eps` choice while playing with the functionality locally.

niksirbi commented on tests/test_unit/test_transforms.py:
```diff
@@ -279,3 +280,143 @@ def verify_log_entry(entry, expected_factor, expected_space_unit):
+        ),
+    ],
+)
+def test_transform(
```
Great job with updating the tests!

niksirbi reviewed (approved):
Thanks for responding to my comments @animeshsasan.

I think the PR is now in excellent shape and I've also provided my opinion on the 2 questions you had.

I'm pre-approving this PR for merging. Before we merge it, I suggest:
- Renaming the PR to sth like "Add function for computing homography-based transforms". The PR's title will become the commit message on `main` after squash-merging (which is our practice).
- Rebasing this branch to the current `main` (just before merging). I don't expect any merge conflicts, but best to sync with `main` to be sure.
